google-cloud-firestore/samples/add_data.rb

```diff
@@ -72,6 +72,54 @@ def set_document_data_types project_id:, collection_path: "data"
   puts "Set multiple data-type data for the one document in the data collection."
 end
 
+# [START firestore_data_custom_type_definition]
+class City
+  attr_accessor :name, :state, :country, :capital, :population, :regions
+
+  def initialize name, state, country, capital: false, population: 0, regions: []
+    @name = name
+    @state = state
+    @country = country
+    @capital = capital
+    @population = population
+    @regions = regions
+  end
+
+  def self.from_h source
+    city = new source[:name], source[:state], source[:country]
+    city.capital = source[:capital] if source[:capital]
+    city.population = source[:population] if source[:population]
+    city.regions = source[:regions] if source[:regions]
+    city
+  end
+
+  def to_h
+    hsh = {
+      name: name,
+      state: state,
+      country: country
+    }
+    hsh[:capital] = capital if capital
+    hsh[:population] = population if population
+    hsh[:regions] = regions if regions
```

Comment:
I think `population` and `regions` should always be truthy because they get set to defaults. We should probably just set all keys in the hash. (And currently this omits `capital` from the hash if it's `false` which is probably not what we'd want.)

Alternatively, if the point is to illustrate that the system can handle hashes with optional keys omitted, then these guards should probably be comparing against the default values instead.

Reply:
I think the idea is to prevent sending nil values. What do you think of this?

```rb
hsh = {
      name: name,
      state: state,
      country: country,
      capital: capital,
      population: population,
      regions: regions
}.delete_if { |_, v| v.nil? }
```

Reply:
Yeah, that works. Since this is sample code, if we have requirements like that for a custom class, we should probably add comments noting that.

Reply:
I'm unsure of the requirements for this example, so we will have to depend on input from the Firestore team. Meanwhile, I added a comment to explain the behavior.